Enable ThrowMissingRegistrationErrors with GraalVM >= 23.1

zakkak opened the pull request:
"This feature will become available as non-experimental in the next release (GraalVM for Java 22), and will be promoted to the default behavior as the community adopts it." -- @vjovanov

See https://github.com/oracle/graal/issues/5173#issuecomment-1752811101

zakkak committed:
Enable ThrowMissingRegistrationErrors with GraalVM >= 23.1

This feature will become available as non-experimental in the next
release (GraalVM for Java 22), and will be promoted to the default
behavior as the community adopts it.

jerboaa commented:
@zakkak Is there a test run of quarkus native tests with a graalvm for JDK 22 build with this somewhere (in order to gauge the impact of this change)?

zakkak commented:
CI run with latest GraalVM for JDK 22-EA: https://github.com/graalvm/mandrel/actions/runs/6531828310

vjovanov commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:
```diff
@@ -814,9 +814,13 @@ public NativeImageInvokerInfo build() {
                     nativeImageArgs.add("--strict-image-heap");
+                    // This feature will become available as non-experimental in the next release (GraalVM for Java 22,
+                    // GraalVM 24.0.0), and will be promoted to the default behavior as the community adopts it.
+                    addExperimentalVMOption(nativeImageArgs, "-H:ThrowMissingRegistrationErrors=");
```
I am afraid this change could break downstream code. I would put a package filter that handles only Quarkus classes and Quarkus-managed libraries in code that goes to users. Then, we can ask the users in the next release to add `--throw-missing-registration-errors=` to their build.

We can use the unfiltered version for internal tests (ideally used with our dev builds). In tests, it is ideally used in combination with `-H:MissingRegistrationReportingMode=Exit` to make sure metadata is never missed.

zakkak replied:
Thanks for the input @vjovanov.

FYI when running without `-H:MissingRegistrationReportingMode=Exit` the Quarkus binary exits with error code 1 without printing anything.

When adding `-H:MissingRegistrationReportingMode=Exit` I get to see the exception:

```
org.graalvm.nativeimage.MissingReflectionRegistrationError: The program tried to reflectively access method sun.security.provider.NativePRNG#<init>(java.security.SecureRandomParameters) without it being registered for runtime reflection. Add sun.security.provider.NativePRNG#<init>(java.security.SecureRandomParameters) to the reflection metadata to solve this problem. See https://www.graalvm.org/latest/reference-manual/native-image/metadata/#reflection for help.
```

My understanding is that Quarkus catches the exception at some point which results in the error message not being shown, yet it still fails.

vjovanov replied:
>My understanding is that Quarkus catches the exception at some point which results in the error message not being shown, yet it still fails.
I would say the same. That is why we introduced `-H:MissingRegistrationReportingMode=Exit`. 
To speed up the migration you can run the programs with `-H:MissingRegistrationReportingMode=Warn` and it will only print the missing elements.

zakkak commented:
Converting to draft due to https://github.com/quarkusio/quarkus/pull/36378#discussion_r1361610015 which requires further investigation.

zakkak committed:
TEST ME DON'T MERGE ME

radcortez committed:
Do not discover properties resource in the classpath to avoid missing resources registration in native mode

cescoffier commented:
@zakkak was the feature added in graalVM/Mandrel?

zakkak commented:
> @zakkak was the feature added in graalVM/Mandrel?

Yes, but Quarkus is still not ready to enable it by default, I created this PR for testing purposes (but I need to rebase it to the latest `main`).
The issue is tracked in https://github.com/quarkusio/quarkus/issues/41995

We are currently looking into solving the issues related to smallrye config (see https://github.com/quarkusio/quarkus/pull/42140 and https://github.com/quarkusio/quarkus/pull/44910 which supersedes https://github.com/quarkusio/quarkus/pull/42140)

zakkak committed:
Catch `MissingReflectionRegistrationError` in loadConfigSources

Works around issue discussed in
https://github.com/oracle/graal/discussions/10264#discussioncomment-11604091

zakkak commented:
Closing as out of date. The issue tracking this is https://github.com/quarkusio/quarkus/issues/41995
